scripts/income_limits/ami.py

```diff
@@ -0,0 +1,192 @@
+"""
+
+# -- 1. AMI data by county/countysub -- #
+# Create table of MFIs and AMIs for each county level
+# (or county subdivision  in the case of New England states)
+
+# Pull MFI, 30%, 50%, 80% AMI from API. Takes ~30 min
+# currently ~10 counties/county subs throw errors, see hud_ami_api_error.log.
+print(f"Pulling AMI data...")
+mfi_ami80_by_countysub = pd.concat([
+    util.pull_state_amis_by_countysub(state_postal_code=state, verbose=True)
+    for state in util.STATE_POSTAL_TO_GEOID.keys()]
+)
+
+# Read in 150% AMI from HAP
+string_cols = ['fips2010', 'State', 'County']
+ami150_by_countysub = pd.read_excel(
+    util.DATA_FPATH / 'raw' / 'il24_all100_150_HAF.xlsx',
+    usecols=['fips2010', 'HAF100_24p4', 'HAF150_24p4'],
+    dtype={'fips2010': str})
+
+ami150_by_countysub.rename(
+    {'fips2010': 'countysub_geoid',
+     'HAF100_24p4': 'ami_100',
+     'HAF150_24p4': 'ami_150'},
+    axis=1,
+    inplace=True)
+
+# Combine amis from the two data sources.
+# Note that the ~10 county/countysubs that we could not pull AMIs for from the API
+# will only have 100% and 150% AMIs since they only appear on the right side of the join
+ami_by_countysub = pd.merge(
+    mfi_ami80_by_countysub,
+    ami150_by_countysub,
+    on='countysub_geoid',
+    how='right',
+    validate="many_to_one")
+
+# add a state column
+geoid_to_postal = {v: k for k, v in util.STATE_POSTAL_TO_GEOID.items()}
+ami_by_countysub['state'] = ami_by_countysub.county_geoid.str.slice(
+    0, 2).map(geoid_to_postal)
+
+# write this now since it takes a while to pull
+
+geo_cols = ["countysub_geoid",  "county_geoid", "state",  "county_name",
+            "counties_msa", "town_name", "metro_status", "metro_name", "area_name"]
+ami_cols = ['median_family_income', 'ami_30',
+            'ami_50', 'ami_80', 'ami_100', 'ami_150']
+
+fpath_out = util.DATA_FPATH / 'processed' / 'ami_by_countysub.csv'
+print(f"Writing to {fpath_out}")
+ami_by_countysub[geo_cols + ['year'] + ami_cols].to_csv(fpath_out, index=False)
+
+# -- 2. Crosswalks from county sub to tract/zcta -- #
+# read in data
+string_cols = ['County code', 'County subdivision (2020)']
+tract_countysub_crosswalk = (
+    util.clean_colnames(pd.read_csv(
+        util.DATA_FPATH / 'raw' / 'geocorr2022_tract_to_countysub.csv',
+        encoding="ISO-8859-1",
+        low_memory=False,
+        skiprows=1,
+        dtype={col: str for col in string_cols}))
+    .drop('county_name', axis=1)
+)
+
+zcta_countysub_crosswalk = (
+    util.clean_colnames(pd.read_csv(
+        util.DATA_FPATH / 'raw' / 'geocorr2022_zcta_to_countysub.csv',
+        encoding="ISO-8859-1",
+        low_memory=False,
+        skiprows=1,
+        dtype={col: str for col in string_cols}))
+    .rename({'zip_census_tabulation_area': 'zcta'}, axis=1)
+    .drop('county_name', axis=1)
+)
+
+# remove non-zctas
+zcta_countysub_crosswalk = zcta_countysub_crosswalk[
+    zcta_countysub_crosswalk.zip_code_name != '[not in a ZCTA]']
+
+# clean up geoids
```

Comment:
Suggestion: Normally I wouldn't ask for this, but since the formatting of the geoids is the whole ballgame here as far as mapping AMIs to the right geography, it might be helpful to have some more documentation as to how exactly you're formatting the geoids. It's not immediately apparent based on the code. 

Alternatively, it's more easily groccable for me if you put `zcta_countysub_crosswalk['county_code'] + zcta_countysub_crosswalk['county_subdivision_2020']`, but you may have performance reasons for structuring the code as you did.